core/src/main/java/tc/oc/pgm/shops/ShopModule.java

```diff
@@ -0,0 +1,191 @@
+  }
+
+  @Override
+  public MatchModule createMatchModule(Match match) {
+    return new ShopMatchModule(match, shops, shopKeepers);
+  }
+
+  @Override
+  public Collection<MapTag> getTags() {
+    return TAGS;
+  }
+
+  public static class Factory implements MapModuleFactory<ShopModule> {
+
+    @Override
+    public ShopModule parse(MapFactory factory, Logger logger, Document doc)
+        throws InvalidXMLException {
+      KitParser kitParser = factory.getKits();
+      PointParser pointParser = new PointParser(factory);
+      Map<String, Shop> shops = Maps.newHashMap();
+      Set<ShopKeeper> keepers = Sets.newHashSet();
+
+      for (Element shop : XMLUtils.flattenElements(doc.getRootElement(), "shops")) {
+        String shopId = XMLUtils.getRequiredAttribute(shop, "id").getValue();
+        String shopName = XMLUtils.getNullableAttribute(shop, "name");
+        List<Category> categories = Lists.newArrayList();
+
+        for (Element category : XMLUtils.getChildren(shop, "category")) {
+          Attribute categoryId = XMLUtils.getRequiredAttribute(category, "id");
+          ItemStack categoryIcon = applyItemFlags(kitParser.parseItem(category, false));
+          List<Icon> icons = parseIcons(category, kitParser, logger);
+          categories.add(new Category(categoryId.getValue(), categoryIcon, icons));
+        }
+
+        if (categories.isEmpty()) {
+          throw new InvalidXMLException("At least one <category> is required per shop", shop);
+        }
+
+        Shop shopInstance = new Shop(shopId, shopName, categories);
+        factory.getFeatures().addFeature(shop, shopInstance);
+        shops.put(shopId, shopInstance);
+      }
+
+      keepers = parseShopKeepers(shops, doc.getRootElement(), pointParser, logger);
+
+      return shops.isEmpty() ? null : new ShopModule(shops, keepers);
+    }
+  }
+
+  private static Set<ShopKeeper> parseShopKeepers(
+      Map<String, Shop> shops, Element root, PointParser pointParser, Logger logger)
+      throws InvalidXMLException {
+    Set<ShopKeeper> keepers = Sets.newHashSet();
+
+    for (Element shopkeeper : XMLUtils.flattenElements(root, "shopkeepers")) {
+      Attribute shopAttr = XMLUtils.getRequiredAttribute(shopkeeper, "shop");
+
+      String shopId = shopAttr.getValue();
+      String name = XMLUtils.getNullableAttribute(shopkeeper, "name");
+      Class<? extends Entity> mob =
+          XMLUtils.parseEntityTypeAttribute(shopkeeper, "mob", Villager.class);
+      ImmutableList<PointProvider> location =
+          ImmutableList.copyOf(pointParser.parse(shopkeeper, new PointProviderAttributes()));
+
+      if (location.isEmpty()) {
+        throw new InvalidXMLException("shopkeeper must have a location defined", shopkeeper);
+      }
+
+      if (!shops.containsKey(shopId)) {
+        throw new InvalidXMLException(
+            "No shop with id '" + shopId + "' could be found", shopkeeper);
+      }
+      Shop shop = shops.get(shopId);
```

Comment:
```suggestion
      Shop shop = shops.get(shopId);
      if (shop == null) {
        throw new InvalidXMLException(
            "No shop with id '" + shopId + "' could be found", shopkeeper);
      }
```
no need to lookup the map twice